distribution/docker/docker-entrypoint.sh

```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env sh
+set -o errexit
+
+# Prevent core dumps
+ulimit -c 0
+
+case $1 in -*) set -- vector "$@"; esac
+
+if [ "$1" = 'help' ]; then
+    # This needs a special case because there's no help output.
+    set -- vector "$@"
+elif vector "$1" --help 2>&1 | grep -q "vector $1"; then
+  set -- vector "$@"
+fi
+
+# If the command is Vector, ensure we run as the `vector` user.
+if [ "$1" = 'vector' ]; then
+  if [ -z "$SKIP_CHOWN" ]; then
+    if [ "$(stat -c %u /etc/vector)" != "$(id -u vector)" ]; then
+      chown -R vector:vector /etc/vector || echo "Could not chown /etc/vector (may not have appropriate permissions)"
+    fi
+
+    if [ "$(stat -c %u /var/lib/vector)" != "$(id -u vector)" ]; then
+      chown -R vector:vector /var/lib/vector || echo "Could not chown /var/lib/vector (may not have appropriate permissions)"
+    fi
+  fi
+
+  if [ "$(id -u)" = '0' ]; then
+    set -- gosu vector "$@"
+  fi
+fi
```

Comment:
I'm not sure doing all this in the entrypoint script is a good idea.
You are basically forcing the user to use your default vector user/group in the image.

There are use-cases where the user will want to run Vector under other UID/GID than the default one in the image. This is why the `docker run` command has a `-u/--user` flag, and even in Docker compose files too.

For instance, I'm pretty sure the container will need to run as `root` in most setups to access the `/var/run/docker.sock` socket, for example for the `docker_logs` source. Or any other resources that normally are owned by root or other non-Vector users in the host.

Defaulting Vector to a non-root user has a number of implications that will break existing setups and most users probably will want to migrate slowly by not being forced to it.

Reply:
Yeah, it's definitely getting messy trying to adhere the the [target spec](https://github.com/vectordotdev/vector/blob/master/docs/specs/target.md) that was produced earlier - especially between the expectations of Agent vs Aggregator roles. An Aggregator requires less (if any) of root level access to things and can be locked down more easily, the Agent (as you pointed out) will regularly need access to resources that are normally more locked down.

We've definitely been focused on the Aggregator role recently, and have begun to present that as the default option - but certainly want to continue supporting and growing the Agent role as well.

I'd like to still provide a "secure by default" configuration, but don't have any great ideas on how to easily allow users to bypass - other than a var similar to `SKIP_CHOWN` that's in the entrypoint now.

Reply:
For us the fundamental role is indeed the Agent. My vector-based load-balancer can be indeed seen as Aggregator role. So we definitively would need the container to work for both cases :)

I agree that all roles (and any further roles such as the "unified" I see in the spec) should be as "secure-by-default" as possible. I think the key here is to think on a "permission model" required by each role. Something like you described above: aggregator=more-sandboxed, agent=more-privileged.

In that case, then definitively the entrypoint script would need to implement that sort of "preparation-before-running".

Given that is impossible to "guess" what role the user wants, the environment variable idea is not that bad actually considering what you exposed. But maybe instead of a low-level `SKIP_CHOWN` variable we could think of a higher-level `VECTOR_ROLE=agent|aggregator|...` one. This could default to `agent` to remain backwards-compatible with previous images. Then the entrypoint script can configure the environment accordingly.

In terms of implementation, the entrypoint could look quite elegant and maintanable like this:
```sh
# prepare running environment according to configured role
case "${VECTOR_ROLE:-agent}" in
  agent) set_agent_env ;;
  aggregator) set_aggregator_env ;;
  *) fatal "unknown role in VECTOR_ROLE, please use 'agent' or 'aggregator'" ;;
esac
```

Such an approach could later be extended easily for future roles.
How does that sound?

Reply:
Dropped the changes to user/group.